### PR TITLE
Move SMT tests to use isoltest

### PR DESCRIPTION
**Pull request by axic**

### CODE REVIEW

**Comment by leonardoalt** [comment]
The `div` test is failing because it expected the solver to say unknown but got SAT. Is the CI using CVC4 as well?

**Comment by leonardoalt** [comment]
Do you think the following is a good idea: one must have both solvers installed in order to update smt tests.
This way we could ensure the most precise result and have CI do the same.

**Comment by axic** [comment]
> The div test is failing because it expected the solver to say unknown but got SAT. Is the CI using CVC4 as well?

I didn't finish this yesterday, but actually was thinking to keep those unknown ones in the C++ code for the moment.

**Comment by ekpyron** [comment]
I rebased this and resolved some conflicts with the JSON AST test extraction.

**Review by ekpyron — DISMISSED** [review]

**Comment by ekpyron** [comment]
Interesting - ``smtCheckerTests/049_division.sol`` takes a long time for me to run (I locally disabled CVC4, since there's still compilation issues on Archlinux) and returns ``UNKNOWN``, hence yielding ``Overflow ... might happen here``, whereas on travis it returns ``SAT``, yielding ``Overflow ... happens here`` - I guess on ~~travis~~ circleci CVC4 is enabled?

We should investigate this. In the long term we will probably have to add expectations that are specific to the enabled solvers similarly to EVM version specific tests in #4262.

I think I will create a ``test system overhaul`` project one of these days and structure all test related issues and improvement suggestions into clean issues and add them to that project (it's not too urgent, though, I think we won't do any of that before ``0.5.1``).

**Comment by ekpyron** [comment]
The same test does seem to return ``UNKNOWN`` on MacOS as well, though.

**Comment by chriseth** [comment]
In general I'm a bit uneasy about this change. The SMT component is not predictable. We might even get test failures depending on the CPU capacity we have because of timeouts etc.

**Comment by chriseth** [comment]
Ok sorry, perhaps put differently: This is not about this PR, but in general, we should be careful about SMT tests.

**Comment by ekpyron** [comment]
@chriseth That's true, but (as you noted) not really related to this PR. I created #4735 for tracking this, but I would say we should still find a more immediate solution for this PR.

File: test/libsolidity/SMTChecker.cpp [hunk 1]
```diff
@@ -1,759 +0,0 @@
...
-	CHECK_SUCCESS_NO_WARNINGS(text);
-	text = R"(
-		contract C {
-			function f(int x, int y) public pure returns (int) {
-				require(y != 0);
-				return x / y;
-			}
-		}
-	)";
-	CHECK_WARNING(text, "Overflow");
```
Comment (ekpyron) [inline test/libsolidity/SMTChecker.cpp]: Ok, the reason why this test didn't fail, resp. behave erratically, before was that we didn't even check whether it was ``Overflow ... might happen`` or ``Overflow ... happens`` before.

**Comment by ekpyron** [comment]
Maybe the easiest (though somewhat unsatisfactory) solution to this is *not* to extract this one test for now and to think of a better solution later on?

**Comment by chriseth** [comment]
What is the state of this PR?

**Comment by leonardoalt** [comment]
I think we can follow @ekpyron 's last suggestion and merge most of the tests (except the ones that use non-linear arithmetics).

**Comment by leonardoalt** [comment]
Rebased and updated this PR to not extract division tests.

**Review by ekpyron — DISMISSED** [review]
The numbering of the extracted tests is somewhat inconsistent (sometimes global prefixes, sometimes per-test suffixes), but probably doesn't matter too much and otherwise: looks fine.

**Comment by leonardoalt** [comment]
Yea I'll submit another PR fixing that and organizing the files in dirs

**Review by chriseth — CHANGES_REQUESTED** [review]
Could you cleanup the commits, please? For me it looks like some back-and-forth in isoltest.cpp.

**Comment by leonardoalt** [comment]
Updated @chriseth

File: test/libsolidity/smtCheckerTests/warn_on_struct.sol [hunk 2]
```diff
@@ -0,0 +1,12 @@
+pragma experimental SMTChecker;
+pragma experimental ABIEncoderV2;
```
Comment (erak) [inline test/libsolidity/smtCheckerTests/warn_on_struct.sol]: Could this be changed, such that it is independent from the ABIEncoderV2? I'd be in favor of trying to factor it out. The following contract seems to trigger the same warnings:
```
pragma experimental SMTChecker;
contract C {
    struct A { uint a; uint b; }
    function f() public pure returns (uint) {
        A memory a = A({ a: 1, b: 2 });
    }
}
```

**Review by erak — COMMENTED** [review]
In general this looks good.

File: test/libsolidity/smtCheckerTests/warn_on_struct.sol (hunk 2 above)
Comment (leonardoalt) [reply test/libsolidity/smtCheckerTests/warn_on_struct.sol]: Makes sense!

Comment (leonardoalt) [reply test/libsolidity/smtCheckerTests/warn_on_struct.sol]: The CI tests actually fail with `TypeError: (116-124): This type is only supported in the new experimental ABI encoder. Use "pragma experimental ABIEncoderV2;" to enable the feature.` without that line. I guess we do need it here.

Comment (erak) [reply test/libsolidity/smtCheckerTests/warn_on_struct.sol]: Does it still fail even with the changes I've proposed? For me it works without ``pragma experimental SMTChecker;`` and it should still test the same.

Comment (leonardoalt) [reply test/libsolidity/smtCheckerTests/warn_on_struct.sol]: To me it works outside the tests, but when running the tests it fails without pragma

**Review by chriseth — DISMISSED** [review]

**Review by erak — DISMISSED** [review]

**Comment by leonardoalt** [comment]
@axic Can we merge this?

**Review by ekpyron — DISMISSED** [review]

**Review by leonardoalt — APPROVED** [review]

**Pull request merged**